pisa/core/binning.py

```diff
@@ -3043,6 +3040,88 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
 
+class VarBinning(object):
+    # pylint: disable=line-too-long
+    r"""
+    Variable binning.
+
+    """
+    # pylint: enable=line-too-long
+    def __init__(self, binnings, selections, name=None):
+
+        assert (isinstance(selections, OneDimBinning) or
+                isinstance(selections, list))
+        assert isinstance(binnings, list) and len(binnings) == len(selections)
+        all_equal = True
+        for b in binnings:
+            assert isinstance(b, MultiDimBinning)
+            if isinstance(selections, OneDimBinning):
+                assert selections.name not in b.names
+            else:
+                assert binnings.count(b) == 1, 'Binning used more than once, modify your selection'
```

Comment:
Do we need to restrict user to use different MultidimBinning for each selection? I feel like it would be more convenient to allow using duplicate binning. For example one can think of a situation where we have three PID bins, but only middle bin has different binning. If one can not use duplicate binning, they will have to define a selection for lower+higher PID bins and then a 3d binning in (E, coszen, PID). 
This of course would not affect the calculation, but, in my opinion, it could just be more convenient to be able to assign same binning to different selections.

Reply:
I only enforce this restriction for arbitrary selections defined in a list of strings. You can use the OneDimBinning selection option for the case of three PID bins where the middle bin should get a different binning in E and coszen. The reason for not allowing it for the arbitrary selections is that this would split your statistic (MC and real event counts) and increase your uncertainty in each bin. In general, if the variable used for the split is also part of the binning, the OneDimBinning option should be used.
Or do you want to have other cuts additional to the PID split that are different for each PID bin?

Reply:
Yes, this was just an example. I agree that having duplicate binning will split your statistic, but if the selection variable carries some information useful for oscillation fit, it could also be balanced out by its contribution to sensitivity. Also like you suggested, it would be useful in the case when there multiple variables used in selection and PID (with different binning for middle bin) is just one of them

Reply:
Hmm, how about making it a warning instead? I think we should also warn that the selection variables should not be part of the MultiDimBinning(s), because the selection itself already effectively acts as binning here

Reply:
Yeah, I think warning would be better. And I think it's a good idea to have a warning about not mixing selection and binding variables.

I also just thought that the same statistic split can be achieved if a variable which has no impact on the fit is used in regular multidim binning. I don't think there is a way to have a warning there though, so it is up to the analyzer to make sure binning variables carry some information

Reply:
yeah, in the end, we have to rely on the analyzers to know what they are doing. :sweat_smile: